api.md

```diff
@@ -0,0 +1,779 @@
+<div align="center">
+
+# `REST`ful API integration
+
+</div>
+
+This guide demonstrates
+how to *extend* our MVP `Phoenix` application
+so it also acts as an **API** 
+returning `JSON` data.
+
+We want our users to securely query
+and manipulate their data 
+and want to ensure all actions 
+that are performed in the Web API
+can also be done through our `RESTful` API
+*and* `WebSocket API`
+(for all real-time updates).
+
+Let's get cracking! 🎉
+
+## 1. Add `/api` scope and pipeline
+
+We want all our API requests
+to be made under the `/api` route.
+This is easier for us to manage changes to API
+that don't create unnecessary regressions to our liveviews.
+
+Let's start by opening `lib/router.ex` 
+and create a new `:api` pipeline
+to be used under `scope "/api"`.
+
+```elixir
+
+  pipeline :api do
+    plug :accepts, ["json"]
+    plug :fetch_session
+  end
+
+  pipeline :authOptional do
+    plug(AuthPlugOptional)
+  end
+
+  scope "/api", AppWeb do
+    pipe_through [:api, :authOptional]
+
+    resources "/items", API.ItemController, only: [:create, :update, :show]
+    resources "/items/:item_id/timers", API.TimerController, only: [:create, :update, :show, :index]
+  end
+```
+
+We are creating an `:api` pipeline 
+that will only accepts and returns `json` objects.
+`:fetch_session` is added as a plug 
+because `:authOptional` requires us to do so.
+
+Every request that is routed to `/api`
+will be piped through both `:api` and `:authOptional` pipelines.
+
+You might have noticed two new controllers:
+`API.ItemController` and `API.TimerController`.
+We are going to need to create these to handle our requests!
+
+## 2. `ItemController` and `TimerController`
+
+Before creating our controller, let's define our requirements. We want the API to:
+
+- read contents of an `item`/`timer`
+- list `timers` of an `item`
+- create an `item` and return only the created `id`.
+- edit an `item`
+
+We want each endpoint to respond appropriately if any data is invalid, 
+the response body and status should inform the user what went wrong. 
+We can leverage changesets to validate the `item` and `timer`
+and check if it's correctly formatted.
+
+### 2.1 Adding tests
+
+Let's approach this
+with a [`TDD mindset`](https://github.com/dwyl/learn-tdd)
+and create our tests first!
+
+Create two new files:
+- `test/app_web/api/item_controller_test.exs`
+- `test/app_web/api/timer_controller_test.exs`
+
+Before implementing,
+we recommend giving a look at [`learn-api-design`](https://github.com/dwyl/learn-api-design),
+we are going to be using some tips coming from there!
+
+We want the API requests to be handled gracefully 
+when an error occurs.
+The person using the API 
+[**should be shown meaningful errors**](https://github.com/dwyl/learn-api-design/blob/revamp/README.md#show-meaningful-errors).
```

Comment:
Let's get https://github.com/dwyl/learn-api-design/pull/47 reviewed/updated & merged _now_ so that we can update this link. 🔗

Reply:
https://github.com/dwyl/learn-api-design/pull/47 merged. ✅